### ISSUE TITLE
Android GPU Service: Occasionally producing bogus vsync timing information

### ISSUE THREAD

**Issue by j9brown** [domokit/mojo#758]
Earlier this week, I noticed that if I launch an app in Mojo Shell on a Nexus 6, occasionally the compositor will crash in gpu_rasterizer.cc due to having received bad vsync timing information from native viewport.  This can be triggered occasionally by repeatedly putting the app in the background then foreground so the GL context gets recreated.  (Note that other crashes can happen due to bugs elsewhere in the stack dealing with GL context loss, particularly a crash down Skia's GPU layer.)

Ok, so the compositor shouldn't be crashing when it gets bad data (I will probably change this so it shuts down the renderer gracefully) but bad vsync timing information is a red flag for possible jank issues so it seems worth investigating

Basically if you print the parameters seen in OnVsyncParametersUpdated, you'll occasionally notice some really odd values.  The crashes I've seen involve vsync intervals on the order of 269 microseconds, which is clearly wrong.

Theory goes that if native viewport is providing bad timing information then that might explain some of the jank seen when running an app on Mojo Shell.  Since I'm out sick today, I recommend one of the owners of Mo'droid look at it and figure out where the bad data is coming from.

**Comment by qsr** [comment]
Will look at this tomorrow, if no one takes it beforehand.

While I'm at it, @j9brown -> what is the expected behavior of the vsync provider. Right now, it is sending vsync information each frame, which mean there is a thread sending a message every 16ms. Is that the expected behavior, or should the information only be send on some other conditions?

**Comment by j9brown** [comment]
The vsync provider should be notified whenever the vsync parameters drift
"too far". Ideally if our system was all tuned and ship-shape then "too
far" might be on the order of 500 us.  Keep in mind that the drift is
cumulative.  So if the vsync interval is off by 100 us then you would need
to send a correction to the vsync timebase every 5 frames or so.

This is something we might want to redesign later.  Note that Android
deliberately uses a polling model for vsync to limit the impact of runaway
apps.  If the GPU service is constantly subscribed to the Android
Choreographer events then it essentially qualifies as a runaway app.  For
now, it's probably enough to say that we only do this when there is an
active GL context (essentially means something is in the foreground).
Later on when we start optimizing for power, we may want to have a way to
turn on/off vsync updates.

**Comment by tonygentilcore** [comment]
Wonky vsync matches some of the traces I've been looking at. This experiment might confirm:
1. @apwilson has an example handy that just moves a red box up and down the screen. Build and run it with a custom `flutter.mojo` engine. Verify that the odd jank repros.
1. Hack `AwaitVSync` to always return `false` here: https://github.com/flutter/engine/blob/master/sky/shell/ui/animator.cc#L116 . If this theory is correct, then we should see a cyclical, but consistent jank rather than the current huge janks.

/cc @abarth @alhaad @apwilson @schwa423 @pylaligand

**Comment by qsr** [comment]
@tonyg Not sure this is exactly the same bug. The vsync you are speaking about seems to be the one flutter uses, while the one Jeff is speaking about is the one that the mozart compositor uses, and they are not the same.

**Comment by qsr** [comment]
The issue with very bogus vsync is fixed. Given the rarity of the event (and it only happening when foregrounding), I really do not hink that it explains jank.

**Comment by j9brown** [comment]
Where's the patch?  Also, did you measure the phase and drift over time?

**Comment by qsr** [comment]
The patch is the one referenced just before.

Also the code is already handling computing the drift and resending data only when the drift is more than 5%. What could be handled better is that it is listening to all vsync to do the computation, while it should probably try to compute what is the drift and sleep some instead, but I do not think this is a priority right now.
